### ISSUE TITLE
[RFC] Complete synchronous TorchFT integration

### ISSUE THREAD

**Issue by Skuld7451** [pytorch/torchtitan#4208]
# RFC 0003: Complete synchronous TorchFT integration

## Summary

Complete TorchTitan's synchronous TorchFT path for configurable initial
synchronization, one-replica execution, replicated/FSDP gradient hooks,
multi-optimizer dispatch, and parity with capabilities recently added to the
ordinary TorchTitan trainer.

## Problem

The experimental trainer had several gaps that become visible outside its
original FSDP-only configuration:

- Initial cross-replica weight synchronization could not be disabled even when
  replicas were known to initialize identically.
- A one-replica job still installed cross-replica wrappers and hooks.
- Replicate-only data parallelism did not have an explicitly validated gradient
  synchronization path.
- `FSDPModule.set_all_reduce_hook()` rejects modules containing multiple FSDP
  parameter groups, leaving MoE layouts that use per-parameter meshes without a
  cross-replica gradient hook.
- Calling the TorchFT optimizer wrapper through the container could recurse
  through the same public `step()` and `zero_grad()` methods.
- Model overrides and chunked-loss `lm_head` setup present in the ordinary
  trainer were not applied by `FaultTolerantTrainer`.

### Initial synchronization and one-replica bypass

`torchtitan/experiments/torchft/manager.py:94-98` exposes `init_sync`, defaulting
to `True`. It is forwarded to TorchFT at lines 141-149:

```python
self._manager = torchft.Manager(
    pg=pg,
    ...,
    init_sync=config.init_sync,
)
```

Managed cross-replica process groups, gradient hooks, loss synchronization, and
the TorchFT optimizer wrapper are used only when `group_size > 1` at
`manager.py:153`, `manager.py:172`, `manager.py:203`, and
`optimizer.py:86-87`. A single replica
therefore retains ordinary local data-parallel behavior.

The command-line opt-out is documented in
`torchtitan/experiments/torchft/README.md:53`:

```bash
--fault-tolerance.no-init-sync
```

It is safe only when corresponding ranks in all replicas begin with identical
model state.

### Cross-replica gradient hook

Native FSDP communication is scoped to the FSDP mesh inside one replica. It
shards and reduces gradients among that replica's ranks, but it does not
synchronize the independently launched TorchFT replicas. Synchronous
multi-replica training therefore needs an additional reduction over
`replicate_pg`: corresponding ranks average their FSDP gradient shards across
replicas before the optimizer update. Without this hook, every replica can
complete valid local FSDP steps while still applying different gradients and
diverging from the others.

The existing TorchTitan integration already installed this hook for a
conventional FSDP module with one parameter group, so this proposal does not add
basic FSDP support from scratch. It extends the supported boundary to
replicate-only DP and FSDP modules containing multiple parameter groups, and it
fails fast when no supported data-parallel wrapper is found. Multi-group
support is needed by layouts such as MoE with per-parameter meshes, for which
the public `FSDPModule.set_all_reduce_hook()` rejects the module.

`torchtitan/experiments/torchft/manager.py:36-56` handles both single- and
multi-parameter-group FSDP modules. The public API remains the default; the
fixed PyTorch version's private parameter-group state is used only when the
public API rejects a module with multiple groups:

```python
def _set_fsdp_all_reduce_hook(module, hook) -> None:
    state = module._get_fsdp_state()
    param_groups = state._fsdp_param_groups
    if len(param_groups) <= 1:
        module.set_all_reduce_hook(hook)
        return

    if not all(hasattr(group, "_all_reduce_hook") for group in param_groups):
        raise RuntimeError(...)
    for param_group in param_groups:
        param_group._all_reduce_hook = hook
```

`manager.py:172-197` traverses both replicate-only and FSDP models. In the
supported PyTorch version, replicate-only modules inherit from `FSDPModule`, so
the same helper covers both modes. Every parameter group receives the same
TorchFT cross-replica average hook; layouts requiring different reduction
semantics per group are outside this proposal's support boundary.
All groups are validated before any private field is updated, so an incompatible
fixed-version API cannot leave a partially installed hook.

For a multi-replica synchronous job, finding no supported data-parallel wrapper
is an explicit configuration error rather than silently training unsynchronized
replicas.

### Non-recursive optimizer adapter

`torchtitan/experiments/torchft/optimizer.py:29-57` introduces a private adapter
that invokes the base container implementation directly:

```python
class _OptimizerAdapter:
    def __init__(self, owner: "TorchFTOptimizersContainer") -> None:
        self._owner = owner

    def step(self, *args, **kwargs):
        return OptimizersContainer.step(self._owner, *args, **kwargs)

    def zero_grad(self, *args, **kwargs):
        return OptimizersContainer.zero_grad(self._owner, *args, **kwargs)
```

The TorchFT wrapper owns this adapter, not the public container methods. Public
dispatch at `optimizer.py:104-123` selects the wrapper for multi-replica
synchronous mode and the adapter otherwise. Exceptions do not mutate the
dispatch policy. Calling the base methods retains the container's existing
multi-optimizer behavior without re-entering the overridden public methods.

### FaultTolerantTrainer parity

`torchtitan/experiments/torchft/trainer.py:111,256,358-378` invokes model
override handling after the training configuration and wires
`ChunkedLossWrapper` to the correct `lm_head` after model parts are built:

```python
def _update_model_config(self, model_config) -> None:
    model_config.update_from_config(config=self.config)
    if self.config.override.imports:
        apply_overrides(self.config.override, self.config)

def _set_chunked_loss_lm_head(self) -> None:
    if not isinstance(self.loss_fn, ChunkedLossWrapper):
        return
    ...
    self.loss_fn.set_lm_head(lm_head)
    model_part._skip_lm_head = True
```

Without these calls, configurations relying on override imports or chunked loss
can work with the ordinary `Trainer` but fail or execute a different model path
with `FaultTolerantTrainer`.

## Compatibility and dependencies

- `init_sync=True` preserves the prior startup behavior.
- Existing multi-replica FSDP jobs retain TorchFT gradient and optimizer
  synchronization.
- One-replica jobs intentionally bypass cross-replica synchronization overhead.
- The change is independent of the process-group registry and LocalSGD CPU
  staging changes.
- Replicate-only support relies on the inheritance contract in TorchTitan's
  required PyTorch version.
- Multi-parameter-group support relies on `_get_fsdp_state()`,
  `_fsdp_param_groups`, and `_all_reduce_hook` from the required PyTorch
  version. An incompatible private API fails explicitly. The long-term fix is a
  public PyTorch FSDP API for per-group hook registration.

## Validation

The following focused coverage was added:

- `tests/test_manager.py:27-154`: `init_sync`, one-replica bypass,
  replicate/FSDP hooks, public single-group dispatch, identical multi-group
  dispatch, incompatible private APIs, invalid unwrapped models, and loss
  synchronization.
- `tests/test_optimizer.py:40-90`: non-recursive dispatch, exception behavior,
  and direct one-replica dispatch.
- `tests/test_trainer.py:30-88`: override ordering and chunked loss for
  non-pipeline, last-stage, and non-last-stage execution.

The focused files contribute 15 tests (11 manager/optimizer and 4 trainer
tests). The full TorchFT test directory used by CI passed:

```text
pytest torchtitan/experiments/torchft/tests -v
16 passed
```

The additional case is the pre-existing checkpoint test, so CI exercises all
four files rather than maintaining a hand-written file list.
